8294916: Cancelling a request must eventually cause its response body subscriber to be unregistered

dfuch opened the pull request:
When [JDK-8277969](https://bugs.openjdk.org/browse/JDK-8277969) was implemented, a list of outstanding response subscribers was added to `HttpClientImpl`. A body subscriber is added to the list after being created and is removed from the list when it is completed, either successfully or exceptionally.

It appears that in the case where the subscription is cancelled before the subscriber is completed, the subscriber might remain registered in the list forever, or at least until the HttpClient gets garbage collected. This can be easily reproduced using streaming subscribers, such as BodySubscriber::ofInputStream. In the case where the input stream is closed without having read all the bytes, Subscription::cancel will be called. Whether the subscriber gets unregistered or not at that point becomes racy.

Indeed, the reactive stream specification doesn't guarantee whether onComplete or onError will be called or not after a subscriber cancels its subscription. Any cleanup that would have been performed by onComplete/onError might therefore need to be performed when the subscription is cancelled too.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294916](https://bugs.openjdk.org/browse/JDK-8294916): Cancelling a request must eventually cause its response body subscriber to be unregistered


### Reviewers
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**) ⚠️ Review applies to [984e7e1d](https://git.openjdk.org/jdk/pull/10659/files/984e7e1dfac8ca1850efed20554841ebf3616fc1)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) ⚠️ Review applies to [984e7e1d](https://git.openjdk.org/jdk/pull/10659/files/984e7e1dfac8ca1850efed20554841ebf3616fc1)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10659/head:pull/10659` \
`$ git checkout pull/10659`

Update a local copy of the PR: \
`$ git checkout pull/10659` \
`$ git pull https://git.openjdk.org/jdk pull/10659/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10659`

View PR using the GUI difftool: \
`$ git pr show -t 10659`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10659.diff">https://git.openjdk.org/jdk/pull/10659.diff</a>

</details>

ashr123 commented:
@dfuch where is the part that removes the subscriber in case `HttpResponse.BodyHandlers#ofInputStream` is used and the `inputStream` closed before the response body is received?

i.e. I expect that even if I use `HttpResponse.BodyHandlers#ofInputStream` with `HttpClient#send` in `while(true)` loop it will retain only 1 subscriber at all times.

Example code for clarification (not complete code):
```java
private static final HttpClient HTTP_CLIENT = HttpClient.newHttpClient();

public static void main(String[] args) throws IOException, InterruptedException {
	final URI uri = URI.create("https://www.oref.org.il/WarningMessages/alert/alerts.json");
	final int minRedAlertEventContentLength = """
			{"cat":"1","data":[],"desc":"","id":0,"title":""}""".getBytes(StandardCharsets.UTF_8).length;

	while (true) {
		final HttpResponse<InputStream> httpResponse = HTTP_CLIENT.send(
				HttpRequest.newBuilder(uri)
						.header("Accept", "application/json")
						.header("X-Requested-With", "XMLHttpRequest")
						.header("Referer", "https://www.oref.org.il/12481-he/Pakar.aspx")
						.timeout(Duration.ofSeconds(10))
						.build(),
				HttpResponse.BodyHandlers.ofInputStream()
		);

		try (InputStream httpResponseBody = httpResponse.body()) {
			if (httpResponse.statusCode() != HttpURLConnection.HTTP_OK) {
				sleep();
				continue;
			}
			if (httpResponse.headers().firstValueAsLong("Content-Length").orElse(-1) > minRedAlertEventContentLength)
				System.out.println(JSON_MAPPER.readValue(
						httpResponseBody,
						RedAlertEvent.class
				));
		}
	}
}

private record RedAlertEvent(
		int cat,
		List<String> data,
		String desc,
		long id,
		String title
) {
}
```

ashr123 commented:
@dfuch see
```java
if (httpResponse.headers().firstValueAsLong("Content-Length").orElse(-1) > minRedAlertEventContentLength)
```
there are cases that the response code is 200 but for my reasons I don't want to read the body (for example body that contains only BOM bytes).

I guess that fixes it too 😃

jaikiran commented on src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java:
```diff
@@ -61,12 +62,28 @@ public void cancel() { }
+
+        @Override
+        public void cancel() {
+            HttpBodySubscriberWrapper.this.cancel(subscription);
```
Hello Daniel, at first when I noticed this code, it looked odd to me that we aren't calling the underlying `subscription.cancel()`. I then noticed that the `HttpBodySubscriberWrapper.cancel()` actually does the cancellation of this underyling `subscription`. It's slightly odd to me that we are doing it there instead of here. Furthermore, the `HttpBodySubscriberWrapper` (which implements the `Flow` interface) has APIs like `onComplete()`, `onError()`. Perhaps it would be more consistent to have an `onCancel` in `HttpBodySubscriberWrapper` and then have the actual cancellation of the `subscription` done here in the `SubscriptionWrapper` and then call the `onCancel(Subscription)` on the `HttpBodySubscriberWrapper`? Something like:



```
class SubscriptionWrapper ... {
  @Override
  public void cancel() {
    try {
      subscription.cancel();
    } finally {
         HttpBodySubscriberWrapper.this.onCancel(subscription);
    }
  }
}
```

```
class HttpBodySubscriberWrapper {
    protected void onCancel(Subscription cancelledSubscription) {
        
    }
}

dfuch replied:
good idea. done.

jaikiran commented on src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java:
```diff
@@ -137,12 +158,12 @@ public CompletionStage<T> getBody() {
             if (subscribed.compareAndSet(false, true)) {
-                userSubscriber.onSubscribe(subscription);
+                SubscriptionWrapper wrapped = new SubscriptionWrapper(subscription);
+                userSubscriber.onSubscribe(this.subscription = wrapped);
```
To make it simpler to read, perhaps change it to:
```
this.subscription = new SubscriptionWrapper(subscription);
userSubscriber.onSubscribe(this.subscription);
```

It's OK with me if you prefer to have it in the current form.

jaikiran replied:
I overlooked the fact that this was a `volatile` field. So this isn't just a stylistic preference and the current form you have here adds value. So yes, looks fine to me. 
I additionally checked the generated bytecode to be sure that this does indeed do just one access to the volatile field (I wasn't sure what it would look like when you are writing to it as well as passing it as a method parameter).

jaikiran commented on src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java:
```diff
@@ -157,7 +178,7 @@ public void onSubscribe(Flow.Subscription subscription) {
         if (completed.get()) {
             if (subscription != null) {
-                subscription.cancel();
+                subscription.subscription.cancel();
```
Are we intentionally accessing and cancelling the underlying subscription here, instead of the wrapper? Do we intentionally want to bypass the on-cancellation logic in this flow?

jaikiran replied:
Thank you for that explanation. Looks good to me.

jaikiran commented on src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java:
```diff
@@ -61,12 +62,28 @@ public void cancel() { }
         this.userSubscriber = userSubscriber;
     }
 
+    class SubscriptionWrapper implements Subscription {
```
Perhaps this can be made `private` class?

jaikiran commented on test/jdk/java/net/httpclient/CancelStreamedBodyTest.java:
```diff
@@ -0,0 +1,432 @@
+    public Object[][] alltests() {
+        String[] uris = uris();
+        Object[][] result = new Object[uris.length * 2][];
+        //Object[][] result = new Object[uris.length][];
```
Leftover comment?

jaikiran commented on test/jdk/java/net/httpclient/CancelStreamedBodyTest.java:
```diff
@@ -0,0 +1,432 @@
+        //Object[][] result = new Object[uris.length][];
+        int i = 0;
+        for (boolean sameClient : List.of(false, true)) {
+            //if (!sameClient) continue;
```
Same here

jaikiran commented on test/jdk/java/net/httpclient/CancelStreamedBodyTest.java:
```diff
@@ -0,0 +1,432 @@
+                .executor(executor)
+                .sslContext(sslContext)
+                .build();
+        return shared ? client : TRACKER.track(client);
```
The implementation of this method looks a bit odd to me. We seem to be creating a new client and then checking if the caller passed `true` for `shared` and if they did then we are returning a brand new client?

Looking at the usage of this `makeNewClient()` method, we should perhaps not even accept a parameter for this method and instead just return a newly created client?

jaikiran commented on test/jdk/java/net/httpclient/CancelStreamedBodyTest.java:
```diff
@@ -0,0 +1,432 @@
+                try (InputStream is = t.getRequestBody()) {
+                    req = is.readAllBytes();
+                }
+                t.sendResponseHeaders(200, -1); // chunked/variable
```
Unrelated to this PR - but this is extra confusing that our `HttpTestExchange` expects `-1` for chunked responses whereas the actual server `com.sun.net.httpserver.HttpExchange` expects `0` for chunked responses.

jaikiran replied:
> I'm not sure I want to change it ;-)

I wasn't expecting this to be changed :) At this point, it's now become a habit to pay extra attention to the value we pass for this method.

jaikiran commented on test/jdk/java/net/httpclient/CancelStreamedBodyTest.java:
```diff
@@ -0,0 +1,432 @@
+                        } catch (InterruptedException x) {
+                            // OK
+                        }
+                        out.printf("Server wrote %d bytes%n", req.length);
```
I think it would be better to move this printf before the call to sleep, since that's when the write completes.

jaikiran commented on test/jdk/java/net/httpclient/CancelStreamedBodyTest.java:
```diff
@@ -0,0 +1,432 @@
+                try (InputStream is = client.send(req, BodyHandlers.ofInputStream()).body()) {
+                    for (int k = 0; k < j; k++) {
+                        read = is.read();
+                        assertEquals(read, BODY.charAt(k));
```
Hello Daniel, I don't follow this loop. The number of times it is iterating seems to be dependent on a statically decided value of `j` (which can't exceed 2). Each iteration it reads 1 byte and compares that byte with the pre-defined response content. Shouldn't this loop instead be doing `k < BODY.length();`?

jaikiran replied:
Now that I think with a fresher mind, I see what this loop does - it intentionally doesn't read the entire response content and instead closes the stream before the whole response is read, to trigger the original issue.

jaikiran commented on test/jdk/java/net/httpclient/CancelStreamedBodyTest.java:
```diff
@@ -0,0 +1,432 @@
+        HttpClient client = null;
+        uri = uri + "/testInputStream";
+        out.printf("%n%s testInputStream(%s, %b)%n", now(), uri, sameClient);
+        for (int i=0; i< ITERATION_COUNT; i++) {
```
Should we reconsider running each test more than once? Here it looks like we are running 3 * 2 = 6 times for testing this use case.

jaikiran commented on test/jdk/java/net/httpclient/CancelStreamedBodyTest.java:
```diff
@@ -0,0 +1,432 @@
+                    // lets split the response in several chunks...
+                    String msg = (req != null && req.length != 0)
+                            ? new String(req, UTF_8)
+                            : BODY;
```
As far as I can see, none of the test requests send a request body. So perhaps we can simplify this part to just always send this static `BODY` content without these additional checks?

dfuch commented:
I have updated the PR. There's a somewhat larger change in how tests can now track HttpClients. I noticed that `check(gracedelay)` wasn't actually checking that all client selector managers threads had exited (as I thought it did) so I changed it to do that, and the SmallTimeout test started failing, so I had to fix that test to null out references and actually wait until all tasks had been executed.

jaikiran commented on test/jdk/java/net/httpclient/CancelRequestTest.java:
```diff
@@ -377,6 +378,13 @@ public void testGetSendAsync(String uri, boolean sameClient, boolean mayInterrup
+
+            var error = TRACKER.check(1,
+                    (t) -> t.getOutstandingOperations() > 0 || t.getOutstandingSubscribers() > 0,
+                    "subscribers for testInputStream(%s)\n\t step [%s]".formatted(req.uri(), i),
```
Hello Daniel,
Is this a copy/paste error - the reference to `testInputStream` method name? Same with other places that have changed in this file.

dfuch replied:
Gah! Yes - thanks for noticing.

jaikiran commented on test/jdk/java/net/httpclient/SmallTimeout.java:
```diff
@@ -79,6 +83,7 @@ static HttpResult of(HttpRequest request, Throwable t) {
     public static void main(String[] args) throws Exception {
         HttpClient client = HttpClient.newHttpClient();
         ReferenceTracker.INSTANCE.track(client);
+        Reference<HttpClient> reference = new WeakReference<>(client);
```
I don't understand the use of `WeakReference` here. I see that we have a `ReachabilityFence` for the `client` in the `finally` block of this test where we then null out `client`. So, if I'm understanding this right, this `WeakReference` is essentially a no-op. i.e. we probably don't need it because we are anyway holding on to the `client` for the lifetime of this test program?

dfuch replied:
The reference is passed to another thread but should remain alive until that other thread has terminated - which is ensured by waiting for the executor to shutdown. What happens here is that the test failed because the other threads started in this block were keeping the reference alive. I'm using a reference here because passing `client` to a lambda makes `client` final and that prevents me from nulling out the `client` variable before calling `gc`. Using a weak reference solves the issue (since I don't have to null it and it won't keep the client alive).

Michael-Mc-Mahon replied:
I'm still unclear as to why the other threads were keeping the reference alive though?

Michael-Mc-Mahon replied:
Okay, I see the comment above about `ReferenceTracker::check(gracedelay)` Never mind

jaikiran commented on test/jdk/java/net/httpclient/SmallTimeout.java:
```diff
@@ -174,6 +181,9 @@ public static void main(String[] args) throws Exception {
             checkReturn(requests);
 
             executor.shutdownNow();
+            if (!executor.awaitTermination(500, TimeUnit.MILLISECONDS)) {
```
Would this timeout value (whatever value we decide) introduce any potential intermittent failures, especially on CI systems? Would it perhaps be better to just call `executor.close()` after that call to `executor.shutdownNow()`, so that if the tasks really don't complete, then the jtreg test timeout will make it fail with a timeout and we don't have to decide what timeout is a good timeout?

jaikiran commented on src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java:
```diff
@@ -61,12 +62,33 @@ public void cancel() { }
+                onCancel();
+            } catch (Throwable t) {
+                onError(t);
+            }
```
Thank you for making this change. Do you think, we should only call `onError()` if `subscription.cancel()` fails? In its current form above, it will end up calling the `onError()` if the `onCancel()` fails for whatever reason.

jaikiran commented:
Thank you for the changes to the new test `CancelStreamedBodyTest`. They looks fine to me.

Michael-Mc-Mahon commented on src/java.net.http/share/classes/jdk/internal/net/http/HttpClientImpl.java:
```diff
@@ -627,7 +638,7 @@ final long unreference() {
         final long webSocketCount = pendingWebSocketCount.get();
-        if (count == 0 && facade() == null) {
+        if (count == 0 && facadeRef.refersTo(null)) {
             selmgr.wakeupSelector();
```
I notice there is only one remaining reference to the facade() method. It might make sense to get rid of it and replace the call site with `facadeRef.refersTo`

dfuch replied:
The only remaining call I see is here:
```
        return new BuilderImpl(this.facade(), proxySelector);
```
and that call needs to return the facade - we can't replace it with `refersTo`. I'm not seeing any other calls.

dfuch commented:
I have updated the PR.

Michael-Mc-Mahon reviewed (approved):
Looks fine to me.

jaikiran reviewed (approved):
Thank you for the updates, Daniel. The latest version of this PR looks good to me.

dfuch commented:
/integrate

dfuch commented:
Integrated. Thanks for all the reviews!

ashr123 commented:
Thank you @dfuch for the effort both here and on StackOverflow!
